core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+      val resize = update.options.contains(ConcurrentWeakHashSet.UpdateOptions.Resize)
+      if (update.options.contains(ConcurrentWeakHashSet.UpdateOptions.RestructureBefore)) {
+        this.restructureIfNecessary(resize)
+      }
+      val skipIfEmpty = update.options.contains(ConcurrentWeakHashSet.UpdateOptions.SkipIfEmpty)
+      if (skipIfEmpty && this.counter.get() == 0) {
+        return update.execute(null) != ConcurrentWeakHashSet.UpdateResults.None
+      }
+      this.lock()
+      try {
+        val index     = this.getIndex(this.references, hash)
+        val head      = this.references(index)
+        val storedRef = this.findInChain(head, element, hash)
+        update.execute(storedRef) match {
+          case ConcurrentWeakHashSet.UpdateResults.None =>
+            false
+          case ConcurrentWeakHashSet.UpdateResults.AddElement =>
+            val newRef = new Ref[V](hash, element, this.queue, head)
+            this.references(index) = newRef
+            this.counter.incrementAndGet()
+            true
+          case ConcurrentWeakHashSet.UpdateResults.RemoveElement =>
+            var previousRef = null.asInstanceOf[Ref[V]]
+            var currentRef  = storedRef
+            while (currentRef != null) {
+              if (currentRef.get() == element) {
+                this.counter.decrementAndGet()
+                if (previousRef == null) {
+                  this.references(index) = currentRef.nextRef // start chain with next ref
+                } else {
+                  previousRef.nextRef = currentRef.nextRef // skip current ref
+                }
+                currentRef = null
+              } else {
+                previousRef = currentRef
+                currentRef = currentRef.nextRef
+              }
+            }
+            true
+        }
+      } finally {
+        this.unlock()
+        if (update.options.contains(ConcurrentWeakHashSet.UpdateOptions.RestructureAfter)) {
+          this.restructureIfNecessary(resize)
+        }
+      }
+    }
+
+    /**
+     * Check if segment needs to be resized and perform resize if necessary.
+     */
+    def restructureIfNecessary(allowResize: Boolean): Unit = {
+      val currentCount = this.counter.get()
+      val needsResize  = allowResize && (currentCount > 0 && currentCount >= this.resizeThreshold)
+      val ref          = this.queue.poll().asInstanceOf[Ref[V]]
+      if (ref != null || needsResize) {
+        this.restructure(allowResize, ref)
+      }
+    }
+
+    /**
+     * Restructure segment by resizing references array and purging dead
+     * references. Given position in the segment is removed if reference was
+     * queued for cleanup by GC or the reference is null or empty.
+     */
+    private def restructure(allowResize: Boolean, firstRef: Ref[V]): Unit = {
+      this.lock()
+      try {
+        val toPurge =
+          if (firstRef != null) {
+            val refs       = mutable.Set[Ref[V]]()
```

Comment:
It would be best to avoid building a mutable set here as this two-phase process encourages the lock and also will decrease performance due to more allocation.

I see that resizing the array is necessary so perhaps we can put it into an `AtomicReference` so we can resize it safely from multiple threads.

Reply:
I really don't like this Set either, but at first I didn't want to change the reference implementation. I think the possibility of improving this part will depend on how the new segment locking mechanism will look like.

Reply:
I took a look and I personally don't see a better option for that with the way how the structure works atm. If we'd remove this Set-based purge, we'll basically end up with standalone `restructure` call for each reference. Let's say there are 10_000 dead refs in the queue, instead of one call we'd need to make like 10k quite heavy calls. Even calling it somehow from different threads, all of them still targets the same segment of map, so it'd effectively be just a worse bottleneck.

Reply:
I've found a relatively easy fix for that while I was trying to optimize impl with atomic references. The `hash` is now mutable, so it's set to `-1` as an indicator of purged ref.